Add TaskMap and TaskInstance.map_id

uranusjr opened the pull request:
TaskMap is populated by a TaskInstance when it pushes its return value to XCom.

Mostly for feedback for now. The pushing part needs some additional logic from #19965 to discover downstreams that need mapping.

uranusjr committed:
Add TaskMap and TaskInstance.map_id

TaskMap is populated by a TaskInstance when it pushes its return value
to XCom.

uranusjr committed:
Only push task map when needed as dependencies

uranusjr committed:
Fix task_reschedule migration

uranusjr committed:
Fix TaskMap table name

uranusjr committed:
Must re-import TaskGrounp at runtime

uranusjr committed:
Use client-side default

ashb commented on airflow/models/taskmap.py:
```diff
@@ -0,0 +1,71 @@
+    dag_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True)
+    task_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True)
+    run_id = Column(String(ID_LEN, **COLLATION_ARGS), primary_key=True)
+    map_id = Column(Integer, primary_key=True)
```
Maybe we should call this `map_index` -- The `id` suffix to me implies its a FK to something, but this isn't an ID itself, but an index in to some other list.

uranusjr replied:
Changed

ashb commented on tests/providers/jira/operators/test_jira.py:
```diff
@@ -39,8 +38,11 @@
-    def setUp(self):
+class TestJiraOperator:
+    def setup_class(self):
+        clear_db_runs()
```
Are these changes related/needed cos of the change in this PR or something else?

uranusjr replied:
I was debugging issues. These will be rolled back if they end up not needed when the PR goes out of draft state.

ashb commented on airflow/migrations/versions/e655c0453f75_add_taskmap_and_map_id_on_taskinstance.py:
```diff
@@ -0,0 +1,124 @@
+    # Create task_map.
+    op.create_table(
+        "task_map",
+        Column("dag_id", String(250, **COLLATION_ARGS), primary_key=True),
```
`StringID` from airflow.models.base

```suggestion
        Column("dag_id", StringID(), primary_key=True),
```

uranusjr committed:
Use StringID in migration

uranusjr committed:
Add map_index as TI and TR init arg

uranusjr committed:
Record map_index to refresh_from_db test

uranusjr committed:
Spellchecker doesn't think "downstreams" is a noun

uranusjr committed:
Try to make migrations work on MySQL ans MSSQL

On MySQL you must not drop an index "needed by a primary key", so we
need to drop the primary key first. On both MySQL and MSSQL, a column
used by a primary key cannot be nullable (which we already accounted
for, but I forgot to add nullable=False in the migration file).

ashb commented on airflow/migrations/versions/e655c0453f75_add_taskmap_and_map_id_on_taskinstance.py:
```diff
@@ -0,0 +1,120 @@
+        Column("dag_id", StringID(), primary_key=True),
+        Column("task_id", StringID(), primary_key=True),
+        Column("run_id", StringID(), primary_key=True),
+        Column("map_index", Integer, primary_key=True),
```
Not that it matters, but for consistency should we have the same default here as we do on TI's map_index column?

uranusjr replied:
The default on `TI.map_index` is mainly added so we don’t need to change too much existing code (otherwise we’d need to add a ton of `map_index=-1`), but this being a new class, I think I prefer being explicit instead.

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -1632,6 +1632,33 @@ def defer(
+                yield key, child
+                if isinstance(child, TaskGroup):
+                    yield from _walk_group(child)
+
+        for key, child in _walk_group(self.dag.task_group):
+            if key == self.task_id:
+                continue
+            if not isinstance(child, (MappedOperator, MappedTaskGroup)):
+                continue
+            if self.task_id in child.upstream_task_ids:
+                return True
+        return False
```
Why do we walk the entire dag, rather than looking at `self.downstream_list`?

uranusjr replied:
Because `downstream_list` misses (mapped) task groups, I believe.

ashb replied:
Could you add a comment saying why?

(Cos I think fairly soon it will be time to have TaskGroups directly in the graph rather than tacked on and UI only like we have it)

uranusjr replied:
Done

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -2128,6 +2135,28 @@ def set_duration(self) -> None:
         self.log.debug("Task Duration set to %s", self.duration)
 
+    @provide_session
+    def _record_task_map_for_downstreams(self, value: Any, *, session: Session = NEW_SESSION) -> None:
```
```suggestion
    def _record_task_map_for_downstreams(self, value: Any, *, session: Session) -> None:
```

I generally avoid using `provide_session` on internal funcs and favour being explicit about passing the session instead.

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -2128,6 +2135,28 @@ def set_duration(self) -> None:
+            task_id=self.task_id,
+            run_id=self.run_id,
+            map_index=self.map_index,
+        ).delete()
```
If we are doing session.merge we don't _have to_ delete it as SQLA would handle doing the right INSERT/UPDATE itself.

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -2128,6 +2135,28 @@ def set_duration(self) -> None:
+        if not self.task.has_mapped_dependants():
+            return
+        if not isinstance(value, collections.abc.Collection):
+            return  # TODO: Error if the pushed value is not mappable?
```
Yeah, I think this should fail the task.

ashb reviewed (commented):
Migration looks good.

ashb reviewed (commented):
We need more tests of `_record_task_map_for_downstreams` behaviour though please.

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -2128,6 +2135,28 @@ def set_duration(self) -> None:
+            run_id=self.run_id,
+            map_index=self.map_index,
+            length=len(value),
+            keys=(list(value) if isinstance(value, collections.abc.Mapping) else None),
```
Maybe it makes sense to move this logic in to the ctor for `TaskMap`?

uranusjr committed:
Move TaskMap creation logic to TaskMap init

uranusjr committed:
Fail ti if unmappable value is pushed for mapping

uranusjr committed:
Move task map recording call a stack higher

This call actually makes more sense in TI._execute_task(), which calls
xcom_push() for the return value, since we get to remove a somewhat
hackish if check.

uranusjr committed:
Some tests (with a bit mocking)

uranusjr committed:
Move decorator-mapping logic to MappedOperator

This makes it easier to keep the logic in sync with BaseOperator.

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -2128,6 +2138,14 @@ def set_duration(self) -> None:
+            return
+        if not isinstance(value, collections.abc.Collection) or isinstance(value, (bytes, str)):
+            self.log.info("Failing %s for unmappable XCom push %r", self.key, value)
+            raise UnmappableXComPushed(value)
```
I don't think we should include the _whole_ value here -- it could potentially be a huge object/structure/etc. (Either here, or in the exception itself.

uranusjr replied:
Oh good point. Even doing `repr(value)[:100]` can potentially consume a lot of memory, so I guess the only reasonable approach is to not log the value at all. Worst case, the user can manually look up that problematic value in the XCom table.

uranusjr replied:
Changed both logging and exception to only include the variable type instead.

uranusjr committed:
Do not log the full unmappable XCom value

An XCom value is arbitrary and can potentially be a giant blob of data,
which would make the log unreadable. Instead, only log the *type* of the
XCom value instead; since the offending TaskInstance's key is also
logged, the user can always look up the actual value in the XCom storage
if needed.

The exception still contains a reference to the actual value so the
catching frame can access it. Since that value is already in memory
anyway, keeping it a bit longer shouldn't be too big a problem.

uranusjr committed:
Note on has_mapped_dependants implementation

We need to traverse the entire DAG to find mapped decendants because
task groups are not currently recorded in the upstream-downstream lists,
only operators. This should change in the future, but until then we need
to make do.

uranusjr committed:
Add 'unmappable' as a valid word

ashb commented on airflow/models/taskmap.py:
```diff
@@ -0,0 +1,110 @@
+# specific language governing permissions and limitations
+# under the License.
+
+"""Table to store "mapped" task instances (AIP-42)."""
```
```suggestion
"""Table to store information about mapped task instances (AIP-42)."""
```

ashb reviewed (approved):
Doc change but LGTM

uranusjr committed:
Clarify TaskMap costring

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

ashb commented on airflow/migrations/versions/e655c0453f75_add_taskmap_and_map_id_on_taskinstance.py:
```diff
@@ -0,0 +1,120 @@
+    with op.batch_alter_table("task_instance") as batch_op:
+        # I think we always use this name for TaskInstance after 7b2661a43ba3?
+        batch_op.drop_constraint("task_instance_pkey", type_="primary")
+        batch_op.add_column(Column("map_index", Integer, nullable=False, default=-1))
```
@uranusjr Did this need `server_default=-1` instead?  - re https://github.com/apache/airflow/issues/20876

uranusjr replied:
I’m investigating if it works as expected on all engines (if not I’ll just do an `UPDATE SET map_index = -1` after this line)

ashb replied:
Doing it with a server default is much preferred -- as otherwise an UPDATE requires us to re-write/touch every row, but a server default (at least on Postgres) doesn't.
